rcl/src/rcl/log_level.c

```diff
@@ -0,0 +1,166 @@
+        RCL_SET_ERROR_MSG("Error while finalizing log levels due to another error");
+      }
+      return RCL_RET_BAD_ALLOC;
+    }
+    dst->logger_settings[i].level =
+      src->logger_settings[i].level;
+  }
+
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_log_levels_fini(rcl_log_levels_t * log_levels)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(log_levels, RCL_RET_INVALID_ARGUMENT);
+  const rcl_allocator_t * allocator = &log_levels->allocator;
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+  if (log_levels->logger_settings) {
+    for (size_t i = 0; i < log_levels->num_logger_settings; ++i) {
+      allocator->deallocate((void *)log_levels->logger_settings[i].name, allocator->state);
+    }
+    log_levels->num_logger_settings = 0;
+
+    allocator->deallocate(log_levels->logger_settings, allocator->state);
+    log_levels->logger_settings = NULL;
+  }
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_log_levels_shrink_to_size(rcl_log_levels_t * log_levels)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(log_levels, RCL_RET_INVALID_ARGUMENT);
+  rcl_allocator_t * allocator = &log_levels->allocator;
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+  if (0U == log_levels->num_logger_settings) {
+    allocator->deallocate(log_levels->logger_settings, allocator->state);
+    log_levels->logger_settings = NULL;
+    log_levels->capacity_logger_settings = 0;
+  } else if (log_levels->num_logger_settings < log_levels->capacity_logger_settings) {
+    log_levels->logger_settings = rcutils_reallocf(
+      log_levels->logger_settings,
+      sizeof(rcl_logger_setting_t) * log_levels->num_logger_settings, allocator);
+    if (NULL == log_levels->logger_settings) {
+      return RCL_RET_BAD_ALLOC;
+    }
+    log_levels->capacity_logger_settings = log_levels->num_logger_settings;
+  }
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_log_levels_add_logger_setting(
+  rcl_log_levels_t * log_levels, const char * logger_name, rcl_log_severity_t log_level)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(log_levels, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(log_levels->logger_settings, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(logger_name, RCL_RET_INVALID_ARGUMENT);
+
+  // check if there exists a same name logger setting
+  rcl_logger_setting_t * logger_setting = NULL;
+  for (size_t i = 0; i < log_levels->num_logger_settings; ++i) {
+    if (strcmp(log_levels->logger_settings[i].name, logger_name) == 0) {
+      logger_setting = &log_levels->logger_settings[i];
+      if (log_levels->logger_settings[i].level != log_level) {
+        RCUTILS_LOG_WARN_NAMED(
+          ROS_PACKAGE_NAME, "Minimum log level of logger [%s] will be replaced from %d to %d",
+          logger_name, logger_setting->level, log_level);
+      }
+      log_levels->allocator.deallocate((void *)logger_setting->name, log_levels->allocator.state);
+      break;
```

Comment:
why deallocating it?
IMO, it's simpler to override the log level and return

```suggestion
          logger_name, logger_setting->level, log_level);
          logger_setting->level = log_level;
      }
      return;
```

Reply:
Updated